mm/alignment.py

```diff
@@ -413,3 +417,261 @@ def run(self):
+        hyp_alignments.setAllophones(self.hyp_allophone_file.get())
+        if isinstance(hyp_alignments, rasr_cache.FileArchiveBundle):
+            hyp_allophone_map = next(iter(hyp_alignments.archives.values())).allophones
+        else:
+            hyp_allophone_map = hyp_alignments.allophones
+
+        ref_alignments = rasr_cache.open_file_archive(self.ref_alignment_cache.get())
+        ref_alignments.setAllophones(self.ref_allophone_file.get())
+        if isinstance(ref_alignments, rasr_cache.FileArchiveBundle):
+            ref_allophone_map = next(iter(ref_alignments.archives.values())).allophones
+        else:
+            ref_allophone_map = ref_alignments.allophones
+
+        file_list = [tag for tag in hyp_alignments.file_list() if not tag.endswith(".attribs")]
+
+        for idx, hyp_seq_tag in enumerate(file_list, start=1):
+            hyp_word_starts, hyp_word_ends, hyp_seq_length = self._compute_word_boundaries(
+                hyp_alignments,
+                hyp_allophone_map,
+                hyp_seq_tag,
+                self.hyp_silence_phone,
+                self.hyp_upsample_factor,
+            )
+            assert len(hyp_word_starts) == len(
+                hyp_word_ends
+            ), f"Found different number of word starts ({len(hyp_word_starts)}) than word ends ({len(hyp_word_ends)}). Something seems to be broken."
+
+            if self.hyp_seq_tag_transform is not None:
+                ref_seq_tag = self.hyp_seq_tag_transform(hyp_seq_tag)
+            else:
+                ref_seq_tag = hyp_seq_tag
+
+            ref_word_starts, ref_word_ends, ref_seq_length = self._compute_word_boundaries(
+                ref_alignments,
+                ref_allophone_map,
+                ref_seq_tag,
+                self.ref_silence_phone,
+                self.ref_upsample_factor,
+            )
+            assert len(ref_word_starts) == len(
+                ref_word_ends
+            ), f"Found different number of word starts ({len(hyp_word_starts)}) than word ends ({len(hyp_word_ends)}) in reference. Something seems to be broken."
+
+            if len(hyp_word_starts) != len(ref_word_starts):
+                logging.warning(
+                    f"Sequence {hyp_seq_tag} ({idx} / {len(file_list)}:\n    Discarded because the number of words in alignment ({len(hyp_word_starts)}) does not equal the number of words in reference ({len(ref_word_starts)})."
+                )
+                discarded_seqs += 1
+                continue
+
+            # Sometimes different feature extraction or subsampling may produce mismatched lengths that are different by one frame, so cut off at the shorter length
+            shorter_seq_length = min(hyp_seq_length, ref_seq_length)
+            hyp_word_ends[-1] = min(hyp_word_ends[-1], shorter_seq_length)
+            ref_word_ends[-1] = min(ref_word_ends[-1], shorter_seq_length)
+
+            assert all(start < shorter_seq_length for start in hyp_word_starts)
```

Comment:
Would this be equivalent to the four assertions?
```suggestion
            assert hyp_word_ends[-1] <= hyp_word_starts[-1] and hyp_word_ends[-1] < hyp_word_ends[-2]
            assert ref_word_ends[-1] <= ref_word_starts[-1] and ref_word_ends[-1] < ref_word_ends[-2]
```
In principle you're only changing the last item, so as long as the last item still has a relative order with respect to the second-to-last item (and is still logically correct with respect to the word start, in the sense that `start < end`), you're guaranteed that the relative order holds for every other item in the list.

> you're guaranteed that the relative order holds for every other item in the list

That should be what `_compute_word_boundaries` already guarantees: it iterates in an orderly manner for the sequence of allophones, and only inserts in an orderly manner the timestamps in `word_starts`, `word_ends` such that the ordering is guaranteed inside each list.

Reply:
I modified the cut-off logic a bit using the logic with iterating from the back of the lists, now everything should be consistent and the assertions are not needed anymore.